feat: allow queries on the Documents tab to be cancelled (COMPASS-5053)

lerouxb opened the pull request:
The

lerouxb commented on packages/compass-crud/src/components/document-list.less:
```diff
@@ -18,6 +18,46 @@
     }
   }
 
+  &-fetching {
```
css largely copied from schema analyse

lerouxb commented on packages/compass-crud/src/components/toolbar.jsx:
```diff
@@ -29,17 +29,14 @@ class Toolbar extends React.Component {
    * Handle loading the previous page of documents in the table view.
    */
   handlePrevPage() {
-    if (this.props.start - 20 <= 0) {
```
Hardcoded integer removed. (bounds get checked inside getPage)

lerouxb commented on packages/compass-crud/electron/renderer/index.js:
```diff
@@ -105,8 +105,6 @@ const dataService = new DataService(connection);
 appRegistry.emit('data-service-initialized', dataService);
 
 dataService.connect((error, ds) => {
-  if (!error) appRegistry.emit('compass:status:done');
```
This is the only thing I have to double-check. It is the last thing outside compass status that emits an event for compass status. Which is just weird. If we really need it I think we should use `appRegistry.getAction('Status.Actions');` like we do everywhere else.

lerouxb commented on packages/compass-crud/src/components/document-list.jsx:
```diff
@@ -82,6 +92,30 @@ class DocumentList extends React.Component {
+   */
+  renderFetching() {
+    return (
+      <div className="document-list-fetching">
```
Didn't seem too trivial because compass-components would be where that goes and that uses different css and testing stuff. Not that this has tests or very much CSS. I guess I can do that..

lerouxb commented on packages/compass-crud/src/stores/crud-store.js:
```diff
@@ -1109,11 +1125,129 @@ async function countWithHint(dataService, ns, filter, countOptions = {}) {
+    reject(new Error(OPERATION_CANCELLED_MESSAGE));
+  };
+
+  return { cancel, promise };
```
Thanks. I was thinking there has _got_ to be something standard by now, but I didn't know what to search for.

gribnoysup commented:
Was playing around with this branch and noticed that when cancelling the initial data fetch I get CRUD view to look like this:

<img width="1027" alt="image" src="https://user-images.githubusercontent.com/5036933/137875485-3f08ad2e-efb6-42af-adbc-1dbaff75dcdb.png">

Probably something that we can handle better

Anemy commented:
Looks like one of the files from the `mongodb-language-model` package we just [recently removed](https://github.com/mongodb-js/compass/pull/2511) made it in - maybe in a rebase? packages/mongodb-language-model/parser.js

Anemy commented:
Really digging this ux - nice work. 
Ran into this when switching to the table documents view:
<img width="472" alt="Screen Shot 2021-10-19 at 9 40 32 PM" src="https://user-images.githubusercontent.com/1791149/138015816-373c2734-2586-41bb-9eca-54fff8545c33.png">
I think we can check `gridApi` exists before destroying it? Looks like it would be caused by being mounted then unmounted quickly, maybe something deeper.

lerouxb commented:
I rewrote some tests to not test against a timeseries collection because it wasn't actually necessary and added a variable (false for now) for skipping the other one.

lerouxb commented:
@Anemy I can easily reproduce that error on the main branch. I think we should just file a ticket rather than fixing even more stuff in this PR.
